### PR TITLE
A great bundle of messy stuff

### PR DESCRIPTION
**Pull request by NiLuJe**
This started as a random "hey, let's run KOReader under callgrind" experiment to see if there weren't a few low-hanging fruits in hot paths, and it kinda involved in a monstruous thing... :D.

I do intend to rebase and squash in a few more-or-less thematic commits before merge (and keep the messy history around in a -wip branch for posterity's sake, because I love me some messy history ^^).

----

* LVDrawBuf:
  * Modernize a few things to use restricted pointers, a few consts, and C++11 types whenever I actually touched the logic.
  * Tightened a very few hot loops, especially in LVColorDrawBuf (e.g., dropped a few useless (or misplaced, e.g., too late/hot guards).

* LVImg:
  * Ditto (with much less impact, because most of it is deferred to the image libraries)
  * Provided NanoSVG is tweaked (as in KOReader) to provide an up-to-date stbiw header, make its implementation private.
  * Resync QImageScale w/ FBInk for a similar modernization pass.

* LVTinyDom:
  * Update to the latest xxHash version (0.8.0), and make the implementation private.
  * Optionally (will be enabled by default in KOReader) switch to ZSTD to handle chunk compression, because zlib is the absolute worse. This should yield slightly smaller cache files, slightly faster compression speeds, and *drastically* faster decompression speeds. There are more performance gains to be had if I ever find the time to train a dictionary.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/399)

### CODE REVIEW

File: crengine/src/lvimg.cpp [hunk 1]
```diff
@@ -1354,7 +1355,17 @@ class CLZWDecoder
...
+        in_stream_size = 0;
+        in_bit_pos = 0;
+
+        p_out_stream = NULL;
+        out_stream_size = 0;
+
+        clearcode = 0;
+        eoicode = 0;
+        bits = 0;
+        lastadd = 0;
```
Comment (NiLuJe) [inline crengine/src/lvimg.cpp:1368]: That was to silence clang-tidy, FWIW.

**Comment by poire-z** [comment]
Any noticable/measurable performance gain, with the drawing and cache stuff?

**Comment by NiLuJe** [comment]
I shaved a whole millisecond and change off in emulator debug builds ;D. Which actually makes that 50% of the runtime, but, yeah, lol :D.

Mostly nothing for grayscale buffers though. I'll check with a vectorized build later, wher eit might have had a small impact.

----

Not *quite* sure how to bench the cache stuff, though. Especially since I don't really have fancier test-cases than the spec data and a few random books I checked ;p.

In *theory*, decompression should be roughly twice as fast.

**Review by poire-z — APPROVED** [review]
Just skimed thru, no comment; trusting you as always on the drawbuf/image stuff - and on this new ZSTD stuff.

**Review by Frenzie — APPROVED** [review]
Looks mostly fairly meaningless, but also hooray for where it's not! ;-)

**Comment by NiLuJe** [comment]
Yeah, on the whole, this should just be a great bundle of churn for not much effect ^^.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 11 of 16 files at r1, 1 of 1 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/koreader/crengine/399)* status: all files reviewed, 1 unresolved discussion

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/koreader/crengine/399)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

**Comment by NiLuJe** [comment]
History cleaned up, messy individual commits archived in https://github.com/NiLuJe/crengine/tree/cre-blit-cxx11_archive

**Comment by Frenzie** [comment]
You can also (perhaps more visibly) archive by opening a new PR btw.

**Comment by NiLuJe** [comment]
> You can also (perhaps more visibly) archive by opening a new PR btw.

True, but I wanted to keep the review in one place ;).

**Comment by poire-z** [comment]
`LVImg: Ditto` might look fine in the global history, but not so fine when git log lvimg.cpp or git blame.

**Comment by NiLuJe** [comment]
Small git sidebar:

Since the git rebase todo UI is not made to deal with diff-related info, and I kinda *really* needed to see which commits touched which file, `git log -p --name-only --reverse upstream/master..HEAD` was my best friend to have a matching layout with that info the next tab over ;).

**Comment by NiLuJe** [comment]
> `LVImg: Ditto` might look fine in the global history, but not so fine when git log lvimg.cpp or git blame.

Good point, I'd forgotten what I'd written in the previous commit :D. Fixed it :).

**Review by poire-z — APPROVED** [review]
(A bit too many blank lines in the first commit message for my liking, but ok :)

**Comment by NiLuJe** [comment]
That's easy enough to deal with ^^.

**Comment by virxkane** [comment]
Since xxHash is declared as a private implementation, maybe xxhash.h should be in crengine/src?

**Comment by poire-z** [comment]
@NiLuJe : just pulled crengine/master to update my local tree to this - without yet the koreader-base stuff, so no change to CMakefile to use ZSTD.
When building, expecting it would still use the legacy ZLIB stuff (that CoolReader still will), I get these kind of errors:
```
[ 66%] Building CXX object CMakeFiles/crengine.dir/crengine/crengine/src/pdbfmt.cpp.o
crengine/src/pdbfmt.cpp: In member function 'bool PDBFile::zlibUnpack(const lUInt8*, size_t, lUInt8*&, lUInt32&)':
crengine/src/pdbfmt.cpp:416:20: error: 'UNPACK_BUF_SIZE' was not declared in this scope
         lUInt8 tmp[UNPACK_BUF_SIZE]; // 256K buffer for uncompressed data
                    ^~~~~~~~~~~~~~~
crengine/src/pdbfmt.cpp:416:20: note: suggested alternative: 'BASE64_BUF_SIZE'
         lUInt8 tmp[UNPACK_BUF_SIZE]; // 256K buffer for uncompressed data
                    ^~~~~~~~~~~~~~~
                    BASE64_BUF_SIZE
crengine/src/pdbfmt.cpp:418:9: error: 'z_stream' was not declared in this scope
         z_stream z = { 0 };
         ^~~~~~~~
crengine/src/pdbfmt.cpp:418:9: note: suggested alternative: '_stream'
         z_stream z = { 0 };
         ^~~~~~~~
         _stream
```
My bad ? Or did you miss some stuff in the !ZSTD case (that I guess you might not have tested? :)

**Comment by NiLuJe** [comment]
Nope, there's a dumb leftover from before the `zlibUnpack` copy ;p

**Pull request merged**